### PR TITLE
Devel

### PR DESCRIPTION
**Pull request by KhWahba**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,19 +1,26 @@
...
-Examples for using the crazyflie-firmware Python bindings
-
-
-## Building
-
-```
-git clone --recursive git@github.com:IMRCLab/pyCrazyflie.git
-cd crazyflie-firmware
-make bindings_python
-```
```
Comment (whoenig) [inline README.md]: I don't think the building instructions should have been removed?

File: Utilities/utils.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    return np.array([[w1],[w2],[w3]])
+
+def Rx(angle):
+    return np.array([[1, 0, 0],[0,np.cos(angle), np.sin(angle)],[0, -np.sin(angle), np.cos(angle)]])
+
+def Ry(angle):
+    return np.array([[np.cos(angle), 0, -np.sin(angle)],[0, 1, 0], [np.sin(angle), 0, np.cos(angle)]])
+
+def Rz(angle):
+    return np.array([[np.cos(angle), np.sin(angle),0],[-np.sin(angle), np.cos(angle),0],[0, 0, 1]])
```
Comment (whoenig) [inline Utilities/utils.py]: Why is that needed? Can't you use rowan instead?

File: controller.py [hunk 3]
```diff
@@ -1,75 +1,165 @@
...
+    state.position.y = uavState[1]    # m
+    state.position.z = uavState[2]    # m
+    state.velocity.x = uavState[3]    # m/s
+    state.velocity.y = uavState[4]    # m/s
+    state.velocity.z = uavState[5]    # m/s
+    q_curr = np.array(uavState[6:10]).reshape((4,))
+    rpy_state  = to_euler(q_curr,convention='xyz')
+    state.attitude.roll  = np.degrees(rpy_state[0])
+    state.attitude.pitch = np.degrees(-rpy_state[1])
+    state.attitude.yaw   = np.degrees(rpy_state[2])
```
Comment (whoenig) [inline controller.py:71]: Also fill state.attitudeQuaternion, since other controllers use that.

File: controller.py [hunk 4]
```diff
@@ -1,75 +1,165 @@
...
+                          state.velocity.x,state.velocity.y, state.velocity.z, 
+                          q_curr[0],q_curr[1],q_curr[2],q_curr[3], 0,0,0]).reshape((13,))
+    return state,fullState
 
+def stateUavStateError(state, uavState):    
+    state_ = np.array([state.position.x, state.position.y, state.position.z, state.attitude.roll, state.attitude.pitch, state.attitude.yaw])
+    q_curr = np.array(uavState[6:10]).reshape((4,))
+    rpy_state  = to_euler(q_curr,convention='xyz')
+    uavState_ = np.array([uavState[0], uavState[1], uavState[2],rpy_state[0], rpy_state[1], rpy_state[2]])
+    return state_ - uavState_
```
Comment (whoenig) [inline controller.py]: For rotation, this is not a proper distance metric. Rowan has some better metrics for orientation error. Where is this function used?

File: genTrajectory/trajectoryGen.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+fig = plt.figure(figsize=(10,10))
+ax  = fig.add_subplot(autoscale_on=True,projection="3d")
+trajspline = ax.plot(postraj[1,:],postraj[2,:],postraj[3,:],'k',lw=2,label="7th order Spline Trajectory")
+refTraj = ax.plot(data[1,:],data[2,:],data[3,:],'-.r',lw=1,label='Reference Trajectory')
+wayp = ax.plot(pk[0,:],pk[1,:],pk[2,:],'*b',lw=5,label='waypoints')
+ax.legend()
+plt.grid()
+plt.show()
+
+
```
Comment (whoenig) [inline genTrajectory/trajectoryGen.py]: Generating a trajectory should not be part of the simulator itself, I think. This should be a separate repository, similar to https://github.com/whoenig/uav_trajectories.

File: initialize.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import numpy as np
+from rowan import from_matrix, to_matrix 
+from utils import Rx, Ry, Rz
+
+dt = 1e-3
+initPos = np.array([0,0,0.7])
+initR = np.eye(3)
+# initialize Rotation matrix about Roll-Pitch-Yaw
+iangle = [0,0,0]
+initR   = Rz((np.pi/180)*iangle[2]) @ Ry((np.pi/180)*iangle[1]) @ Rx((np.pi/180)*iangle[0])
```
Comment (whoenig) [inline initialize.py]: use rowan and np.radians instead

File: initialize.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+initq = from_matrix(initR)
+#Initialize Twist
+initTwist = np.zeros((6,))
+### State = [x, y, z, xdot, ydot, zdot, q1, q2, q3, q4, wx, wy, wz] ###
+initState = np.zeros((13,))
+initState[0:3]  = initPos  # position: x,y,z
+initState[3:6]  = initTwist[0:3]  # linear velocity: xdot, ydot, zdot
+initState[6:10] = initq# quaternions: [q1, q2, q3, q4]
+initState[10::] = initTwist[3::] # angular velocity: wx, wy, wz
```
Comment (whoenig) [inline initialize.py]: avoid polluting the namespace like that (this can lead to very strange behavior, if you have another module that uses the same variable names). Instead, put this in a function that you call (or put it as part of your main controller.py script)

File: uavDy/uav.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            self.m      = 0.028 
+            self.I      = np.array([[16.571710, 0, 0],[0, 16.655602, 0],[0, 0, 29.261652]])* 1e-6 
+            self.invI   = linalg.inv(self.I)
+            self.d      = 4e-3
+            self.cft    = 0.005964552 
+            self.all    = np.array([[1, 1, 1, 1],[0, -self.d, 0 , self.d],[self.d, 0 , -self.d, 0],[-self.cft, self.cft, -self.cft, self.cft]])
+            self.invAll = linalg.pinv(self.all)
+            self.grav  = np.array([0,0,-self.m*9.81])
+             ### State initialized with the Initial values ###
+             ### state = [x, y, z, xdot, ydot, zdot, q1, q2, q3, q4, wx, wy, wz]
```
Comment (whoenig) [inline uavDy/uav.py]: for quaternions, always try to be specific where the x/y/z and where w is. For rowan, it is qw, qx, qy, qz, but many people use qx, qy, qz, qw instead.

**Review by whoenig — CHANGES_REQUESTED** [review]
Cool that this works now! High-level:

* Try to make this PR as small as possible. In particular, I recommend not including any of the trajectory optimization/generation for now. It is OK to include one already generated example trajectory (such as the figure8)
* Make use of existing libraries whenever possible. In this case, use numpy and rowan, rather than your own implementation (better tested, easier to read for newcomers)

Low-Level: see comments.

You can update your PR by directly pushing new changes to your branch. Once you are ready, just request another review from me via the github webpage.

File: README.md (hunk 1 above)
Comment (KhWahba) [reply README.md]: Yes I don't think so too, I will add it again.

File: Utilities/utils.py (hunk 2 above)
Comment (KhWahba) [reply Utilities/utils.py]: It was used to initialize the roll, pitch and yaw using rotations, but I will remove it and use rowan instead

File: controller.py (hunk 4 above)
Comment (KhWahba) [reply controller.py]: This was just debugging to check that what is passed from the dynamic model states to the controller is the same, I forgot to delete it because it was just for checking but it is misrepresented.

File: initialize.py (hunk 7 above)
Comment (KhWahba) [reply initialize.py]: I removed the full initialize.py and added a function inside the main controller.py that sets the initial value, to avoid this mess.

File: initialize.py (hunk 6 above)
Comment (KhWahba) [reply initialize.py]: Done

File: controller.py (hunk 3 above)
Comment (KhWahba) [reply controller.py:71]: Okay

File: genTrajectory/trajectoryGen.py (hunk 5 above)
Comment (KhWahba) [reply genTrajectory/trajectoryGen.py]: I will remove it and keep the trajectoriescsv/ directory to keep .csv files to load it in the simulator (i.e infinity trajectory to test with it)

File: uavDy/uav.py (hunk 8 above)
Comment (KhWahba) [reply uavDy/uav.py]: Okay, done.

File: README.md [hunk 9]
```diff
@@ -1,19 +1,43 @@
...
-## Building
-
+## Building Instructions
 ```
 git clone --recursive git@github.com:IMRCLab/pyCrazyflie.git
 cd crazyflie-firmware
 make bindings_python
 ```
+---
+---
```
Comment (whoenig) [inline README.md]: What are these lines for?

File: controller.py [hunk 10]
```diff
@@ -1,75 +1,187 @@
 #!/usr/bin/env python
+import sys
+import os
+path = os.path
+sys.path.append('uavDy/')
+sys.path.append('animation/')
+sys.path.append('trajectoriescsv/')
```
Comment (whoenig) [inline controller.py]: not needed anymore

File: controller.py [hunk 11]
```diff
@@ -1,75 +1,187 @@
 #!/usr/bin/env python
+import sys
+import os
+path = os.path
+sys.path.append('uavDy/')
+sys.path.append('animation/')
```
Comment (whoenig) [inline controller.py]: avoid that. Instead it should be possible to use relative imports? (`from animation import ...`)

File: uavDy/uav.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+            curr_vel  = self.state[3:6]  # linear velocity: xdot, ydot, zdot
+            curr_q    = self.state[6:10] # quaternions: [q1, q2, q3, q4]
+            curr_w    = self.state[10::]  # angular velocity: wx, wy, wz
+            
+            posNext, velNext = self.getNextLinearState(curr_vel, curr_pos, curr_q, fz)
+            qNext, wNext     = self.getNextAngularState(curr_w, curr_q, tau_i)
+            
+            self.state[0:3]  = posNext  # position: x,y,z
+            self.state[3:6]  = velNext  # linear velocity: xdot, ydot, zdot
+            self.state[6:10] = qNext# quaternions: [q1, q2, q3, q4]
```
Comment (whoenig) [inline uavDy/uav.py]: update comment to specify which quaternion components are which

**Review by whoenig — COMMENTED** [review]
Looks good overall. Just a few minor comments left.

File: README.md (hunk 9 above)
Comment (KhWahba) [reply README.md]: I removed them

File: controller.py (hunk 11 above)
Comment (KhWahba) [reply controller.py]: Yes it is possible to use relative imports, I will do it. 
I also added in the readme to add the crazyflie-firmware to the PYTHONPATH (as in the main branch readme) and I added a "try: ... except" for the import of cffirmware in the "if __name__ == '__main__':" condition. I think it is better to be added manually, because if someone is using another firmware code in another path.

File: uavDy/uav.py (hunk 12 above)
Comment (KhWahba) [reply uavDy/uav.py]: Done.

**Review by whoenig — APPROVED** [review]
Looks good!
